Dictionaries V1

cesarvr opened the pull request:
## What, How & Why?
<!-- Describe the changes and give some hints to guide your reviewers if possible. -->
<!-- E.g. reference to other repos: This closes realm/realm-sync#??? -->

This closes # ???

## ☑️ ToDos
<!-- Add your own todos here -->
* [x] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [ ] 📝 Public documentation PR created or is not necessary
~~* [ ] 💥 `Breaking` label has been applied or is not necessary~~

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

fronck commented on src/common/deleteme/collection.hpp:
```diff
@@ -0,0 +1,8 @@
+//
```
[Question]  This file is in a directory called "deleteme".  Is it supposed to be checked in?

fronck commented on src/node/node_type_deduction.hpp:
```diff
@@ -52,14 +70,19 @@ struct TypeDeductionImpl {
     }
 
-    static std::string to_javascript(types::Type value)
+    static bool realm_type_exist(std::string type){
```
```suggestion
    static bool realm_type_exist(std::string const &type){
```

fronck replied:
In general, it's good practice to `const` function arguments when possible.  It allows calling functions with `const` arguments to call this function.  It allows you to pass by reference, possibly avoiding a copy.

fronck commented on src/node/node_type_deduction.hpp:
```diff
@@ -52,14 +70,19 @@ struct TypeDeductionImpl {
+        return realm_types.find(type) == realm_types.end();
+    }
+
+    static types::Type realm_type(std::string type){
```
```suggestion
    static types::Type realm_type(std::string const &type){
```

fronck commented on src/test/js_dictionary.cpp:
```diff
@@ -0,0 +1,11 @@
+using Catch::Matchers::Contains;
+
+TEST_CASE( "Testing DictionarySchema#Constructor" ) {
+    realm::js::DictionarySchema schm {"string{}"};
```
[Questions]  Is the postfix for `dictionary` not supposed to be `<>`?  I'm using `{}` for `set` at the moment.

kneth replied:
Our dictionaries look very much like plain JavaScript objects (key is a string). For example `{ name: "Arthur", age: 42 }` - so maybe `{}` is a better fit for dictionaries.

fronck commented on src/test/test.sh:
```diff
@@ -0,0 +1,10 @@
+g++ -std=c++17 -Wall -I catch.hpp -I ../ -I ../../vendor/realm-core/src/realm/object-store/ \
```
Please add some comments about what this script does.  It's not obvious to me at least.
Also, should it be moved to the main `scripts/test.sh` as an parameter?

fronck commented on src/test/test.sh:
```diff
@@ -0,0 +1,10 @@
+    && 
+
+./testing  
+# missing realm::ConstLstIf
```
[Question]  Is this comment relevant here?

fronck commented on src/node/node_type_deduction.hpp:
```diff
@@ -25,8 +29,22 @@
+        std::map<std::string, types::Type> ret;
+        for(auto& [type, key] : realm_types ){
+            ret[key] = type;
+            std::transform(key.begin(), key.end(), key.begin(), [](unsigned char c){ return std::tolower(c); });
```
```suggestion
            std::transform(key.begin(), key.end(), key.begin(), [](static_cast<unsigned char>(c)){ return std::tolower(c); });
```

fronck commented on src/js_schema.hpp:
```diff
@@ -85,6 +85,13 @@ static inline void parse_property_type(StringData object_name, Property& prop, S
         type = type.substr(0, type.size() - 1);
     }
 
+    DictionarySchema dictionary {type};
+
+    if(dictionary.is_dictionary()){
+        prop.type |= dictionary.schema();
+        return;
+    }
```
[Suggestion]  Adding a guard here to make sure that `DictionarySchema dictionary {type}` isn't executed on every invocation of `parse_property_type`.  It's a rather costly operation..

fronck reviewed (changes requested):
:+1:, but a few comments for your consideration.

kneth commented on src/dictionary/methods/listeners.hpp:
```diff
@@ -0,0 +1,77 @@
+
+    template <typename JavascriptPlainObject>
+    auto add_listener(JavascriptPlainObject* object) {
+        return [=](const Napi::CallbackInfo& info) {
```
By using `js::Function<VM>` you can support any engine.

cesarvr replied:
I would like to do it, but that [Napi::CallbackInfo](https://github.com/nodejs/node-addon-api/blob/main/doc/callbackinfo.md) is not a function, is some kind of object injected into functions. I would suggest waiting for the port to JSC and we re-assess.

bmunkholm commented on src/dictionary/methods/callbacks.hpp:
```diff
@@ -0,0 +1,76 @@
+//
+// Created by Cesar Valdez on 07/03/2021.
```
replace w standard header

cesarvr replied:
My name looks nice there, was joking good catch.

bmunkholm commented on src/dictionary/methods/listeners.hpp:
```diff
@@ -0,0 +1,77 @@
+//
+// Created by Cesar Valdez on 07/03/2021.
```
replace with standard header.

bmunkholm commented on src/js_schema.hpp:
```diff
@@ -85,6 +85,13 @@ static inline void parse_property_type(StringData object_name, Property& prop, S
 
+    DictionarySchema dictionary {type};
+
+    if(dictionary.is_dictionary()){
```
Use spacing similar to the rest of the code. After "if", before "{".

kneth replied:
Please ignore code formatting - we plan to apply `clang-format` using Realm Core's configuration when we merge `develop` to `master` (most likely: immediately after the merge).

cesarvr commented:
# Added Support For JavaScript Core (JSC)

## Added
- New methods remove/put. This way users can update / remove keys from their dictionary.
- Object mutation to reflect the contents of the dict.
- Added support for JavaScript Core API (react-native).
- Added Catch2 framework (simple implementation using just the headers).
- Added a [test bed to run our C++ test](https://github.com/realm/realm-js/blob/5264c05d090e6158b33b5bda7e77b2a32c017e5d/src/test/README.md) against JSC.

kneth commented on src/dictionary/methods/accessors.hpp:
```diff
@@ -0,0 +1,59 @@
+    void set(accessor::Arguments args){
+        auto context = args.context;
+        auto key = args.property_name.c_str();
+        try{
```
Is this needed? I mean, won't the `wrap` of `RealmObjectClass<T>::set_property()` catch it (https://github.com/realm/realm-js/blob/master/src/js_realm_object.hpp#L89)?

cesarvr replied:
Yes but it will make the module untestable.

Leaving us two choices tested code or perpetuating a bad practice with the use of global variables.

kneth commented on src/js_object_accessor.hpp:
```diff
@@ -235,7 +242,7 @@ class NativeAccessor {
 
     bool is_same_dictionary(realm::object_store::Dictionary const& dictionary, ValueType const& value) const {
-        throw std::runtime_error("'Dictionary' type support is not implemented yet");
+        return false;
```
Is there no way we can compare to dictionaries?

cesarvr replied:
We can but we don't need to, or at least in the current implementation. Can you suggest a use case for this ?

kneth commented on src/test/catch.cpp:
```diff
@@ -0,0 +1,9430 @@
+//              Copyright Catch2 Authors
```
Is it possible to reuse the `Catch2` version in core: https://github.com/realm/realm-core/tree/master/external? If no, I would prefer to include it as a git submodule.

cesarvr replied:
It's possible, but I didn't have time to read the the extensive documentation in realm-core explaining the way they use it. 

Also pressed by time have to go with something simple and that solve our problem.

kneth commented on src/test/test.sh:
```diff
@@ -0,0 +1,12 @@
+g++ -g -std=c++17 -Wall -I catch.hpp -I test_bed.hpp -I ../ \
```
This seems to be very MacOS specific. Of course, we are mostly using MacOS but I think it would be better to have a proper CMake configuration.

cesarvr replied:
Because it should work for us, edge cases running Linux are welcome to contribute with a PR :).

cesarvr replied:
Because it should worked me, edge cases are welcome to contribute with a PR :).

kneth commented on src/CMakeLists.txt:
```diff
@@ -1,16 +0,0 @@
-add_library(realm-js-shared OBJECT
```
Why is this file deleted?

cesarvr replied:
That was my ide that did that by accident will delete it if not necessary :)

cesarvr commented on src/common/object/jsc/object.hpp:
```diff
@@ -0,0 +1,230 @@
+    std::vector<std::string> accessors;
+    PrivateStore *private_object;
+
+    static std::string to_string(JSContextRef context, JSStringRef value) {
```
Because that one is not reusable, unless you pull the full global variable in the form of template type. 

In order to make this module self contain and testable I/we need to avoid pulling global variables in disguise like the one necessary to reuse the function you mention.

cesarvr commented on src/common/object/node/methods.hpp:
```diff
@@ -0,0 +1,57 @@
+auto NodeCallbackWrapper(const Napi::CallbackInfo &values) {
+    return [&](int index) -> Napi::Value { return values[index]; };
+}
+namespace method {
```
I think templates aren't meant to solve the problem I'm trying to solve here, which is to target/optimise for different architectures. 

I try to keep a homogeneous type across nodejs and JSC.

cesarvr commented on src/common/object/strategies.hpp:
```diff
@@ -0,0 +1,38 @@
+
+#include "common/object/interfaces.hpp"
+
+#pragma once
```
I think that's unnecessary because those modules above should contain the same protection. If they don't we want to know.

cesarvr commented on src/common/object/error_handling.hpp:
```diff
@@ -0,0 +1,47 @@
+////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#if REALM_PLATFORM_NODE
```
Sorry that's a dead module will delete it.

cesarvr commented on src/common/js_plain_object.hpp:
```diff
@@ -0,0 +1,136 @@
+
+    std::unique_ptr<Builder> builder;
+    std::unique_ptr<Collection> collection;
+    std::vector<Subscriber*> subscribers;
```
Good catch, 👍 I would fix that soon.

Would run some instrumentation to catch others lurking in the dark :)

cesarvr commented on src/common/js_plain_object.hpp:
```diff
@@ -0,0 +1,136 @@
+
+    void remove_subscription(const Subscriber* subscriber) {
+        int index = -1;
+        for (auto const& candidate : subscribers) {
```
I try that the problem is that it makes a transformation in place leaving empty spaces and providing the wrong size for the vector at least in my experience.

I would double check this as I'm all in to make sure that's the case.

fronck commented on src/dictionary/methods/functions.hpp:
```diff
@@ -0,0 +1,102 @@
+        IOCollection *collection = arguments.collection;
+
+        object_keys(context, arguments.get(0, error_msg),
+                    [=](std::string& key, auto object){
```
```suggestion
                    [=](std::string& key, auto object) {
```

fronck commented on src/dictionary/methods/subscriber.hpp:
```diff
@@ -0,0 +1,93 @@
+        : fn{_context, _fn},
+          context{Context<T>::get_global_context(_context)} {}
+
+    DictionaryChangesSubscriber(DictionaryChangesSubscriber&& _subscriber){}
+
+    DictionaryChangesSubscriber(DictionaryChangesSubscriber& _subscriber){}
```
```suggestion
    DictionaryChangesSubscriber(DictionaryChangesSubscriber&& _subscriber) {}

    DictionaryChangesSubscriber(DictionaryChangesSubscriber& _subscriber) {}
```

fronck commented on src/dictionary/methods/subscriber.hpp:
```diff
@@ -0,0 +1,93 @@
+    auto build_array(Collection &collection) const {
+        std::vector<ValueType> values;
+        for (auto mixed_item : collection) {
+            values.push_back(
```
```suggestion
            values.emplace_back(
```

fronck commented on src/dictionary/methods/subscriber.hpp:
```diff
@@ -0,0 +1,93 @@
+        Function<T>::callback(context, fn, object, 2,
+                              arguments);
+    }
+    ~DictionaryChangesSubscriber(){}
```
```suggestion
    ~DictionaryChangesSubscriber() {}
```

fronck commented on src/js_types.hpp:
```diff
@@ -24,6 +24,7 @@
 #include <string>
 #include <vector>
 #include <sstream>
+#include <iostream>
```
Is this include required, or a leftover from debugging?

cesarvr replied:
yep :), good catch 👍

fronck commented on src/test/README.md:
```diff
@@ -0,0 +1,142 @@
+# How To Test C++ And JS
+
+Usually to test against Javascript Core (JSC) we have to boot up an Android emulator, with all its dependencies (SDK, NDK, etc.) and then
+go through the suffering of having to create/load a react-native dummy app. Usually this way of working will require at between `20-40G` of space and lots of patience.
```
```suggestion
go through the process of having to create/load a react-native dummy app. Usually this way of working will require at between `20-40G` of space and lots of patience.
```

fronck commented on src/test/README.md:
```suggestion
### APIs
```

fronck commented on src/test/README.md:
```diff
@@ -0,0 +1,142 @@
+Usually to test against Javascript Core (JSC) we have to boot up an Android emulator, with all its dependencies (SDK, NDK, etc.) and then
+go through the suffering of having to create/load a react-native dummy app. Usually this way of working will require at between `20-40G` of space and lots of patience.
+
+To avoid all that pain we should target JSC directly and avoid the expensive middlemen for all code that aims directly to JSC (roughly 90%) and while 
```
```suggestion
To avoid all that complexity we should target JSC directly and avoid the expensive middlemen for all code that aims directly to JSC (roughly 90%) and while 
```

fronck replied:
Isn't this something that your implementation already does?  ("should" makes it sound like we've taken a shortcut in our implementation that we shouldn't have)

cesarvr replied:
Just thought that this addition (Catch3 and the JSC boilerplate) might deserve a justification as to why I think is good to have it.

fronck commented on src/test/README.md:
```suggestion
   *  Previous boilerplate would have been here...
```

fronck reviewed (changes requested):
Please see the inline comments.

On a personal note -- in the interest of improving code quality -- I'd really like to see some Doxygen documentation of, at least, the major functionality.  This would also make the code logic easier to follow during code review.

cesarvr commented on src/common/object/node/methods.hpp:
```diff
@@ -0,0 +1,57 @@
+        Napi::Value value;
+
+        void throw_error(std::string&& message){
+            throw Napi::Error::New(context, message);
```
They are, but only if I use the global variables that will make the code very hard to test. To make this piece of work testable I decided to implement/replicate some functionalities in the form of isolated/testable modules. this is one of those and hopefully we can migrate more functionality into a state that we can test it without requiring an emulator.

kneth reviewed (commented):
As our code base is heavily underdocumented, I suggest that comments like in https://github.com/realm/realm-js/blob/develop/src/js_set.hpp is added.

cesarvr commented on src/common/object/node/object.hpp:
```diff
@@ -0,0 +1,157 @@
+        auto object = get();
+
+        auto method = [=](const auto &info) {
+            callback({info.Env(), observer, collection, info.Length(),
```
> Do we need a HANDLESCOPE here?
No, because the JSObject class shouldn't be not concerned about the current scope that the user (the dev) want to use.
